File: 028.py
#   028
#   Implement strStr()
#   2015-05-03
#####################################################
class Solution:
    # @param {string} haystack
    # @param {string} needle
    # @return {integer}
    def strStr(self, haystack, needle):
        if len(needle) == 0:
            return 0
        for i in range(0, len(haystack)):
            if haystack[i] != needle[0]:
                continue
            k = i
            j = 0
            while j < len(needle):
                if k == len(haystack):
                    return -1
                if haystack[k] == needle[j]:
                    k+=1
                    j+=1
                else:
                    break
            if j == len(needle):
                return i
        return -1

    def strStr0(self, haystack, needle):
        return haystack.find(needle)

    def strStr2(self, haystack, needle):    # KMP
        l2 = len(needle)
        if l2 == 0:
            return 0
        T = [-1]
        for j in range(1, l2):
            if T[j-1] > -1 and needle[j-1] == needle[T[j-1]]:
                T.append(T[j-1] + 1)
            else:
                T.append(0)
        l1 = len(haystack)
        i = 0
        j = 0
        while i + j < l1:
            if haystack[i+j] == needle[j]:
                if j == l2 - 1:
                    return i
                else:
                    j+=1
            else:
                if T[j] > -1:
                    i = i + j - T[j]
                    j = T[j]
                else:
                    i += 1
                    j = 0
        return -1

#####################################################
haystack = "ABC ABCDAB ABCDABCDABDE"
needle = "ABCDABD"

sol = Solution()
print sol.strStr2(haystack, needle)

File: ruby/028.rb
# 028
# Implement strStr()
# 2015-05-03
#####################################################
# @param {String} haystack
# @param {String} needle
# @return {Integer}
def str_str(haystack, needle)
  return 0 if needle.size == 0

  for i in (0...haystack.size)
    if haystack[i] != needle[0]
      next
    end

    k = i
    j = 0
    while j < needle.size
      if k == haystack.size
        return -1
      end

      if haystack[k] == needle[j]
        k+=1
        j+=1
      else
        break
      end
    end
    if j == needle.size
      return i
    end
  end
  -1
end

def str_str0(haystack, needle)
  return haystack.index(needle) || -1
end

def str_str2(haystack, needle)  # KMP
  l2 = needle.size
  if l2 == 0
    return 0
  end
  t = [-1]
  for j in (1...l2)
    if t[j-1] > -1 and needle[j-1] == needle[t[j-1]]
      t[j] = t[j-1] + 1
    else
      t[j] = 0
    end
    puts "#{needle[j]} #{t[j]}"
  end

  l1 = haystack.size
  i = 0
  j = 0
  while i + j < l1
    if haystack[i+j] == needle[j]
      if j == l2 - 1
        return i
      else
        j+=1
      end
    else
      if t[j] > -1
        i = i + j - t[j]
        j = t[j]
      else
        i += 1
        j = 0
      end
    end
  end
  -1
end

#####################################################

haystack = "ABC ABCDAB ABCDABCDABDE"
needle = "ABCDABD"
puts str_str2(haystack, needle)

